src/main/java/org/prebid/server/floors/PriceFloorFetcher.java

```diff
@@ -0,0 +1,316 @@
+import org.prebid.server.util.HttpUtil;
+import org.prebid.server.util.ObjectUtil;
+import org.prebid.server.vertx.http.HttpClient;
+import org.prebid.server.vertx.http.model.HttpClientResponse;
+
+import java.math.BigDecimal;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class PriceFloorFetcher {
+
+    private static final Logger logger = LoggerFactory.getLogger(PriceFloorFetcher.class);
+    private static final ConditionalLogger samplingLogger = new ConditionalLogger(logger);
+    private static final int ACCOUNT_FETCH_TIMEOUT = 10;
+
+    private final ApplicationSettings applicationSettings;
+    private final Metrics metrics;
+    private final Vertx vertx;
+    private final TimeoutFactory timeoutFactory;
+    private final HttpClient httpClient;
+    private final JacksonMapper mapper;
+
+    private final Set<String> fetchInProgress;
+    private final Map<String, AccountFetchContext> fetchedData;
+
+    public PriceFloorFetcher(ApplicationSettings applicationSettings,
+                             Metrics metrics,
+                             Vertx vertx,
+                             TimeoutFactory timeoutFactory,
+                             HttpClient httpClient,
+                             JacksonMapper mapper) {
+        this.applicationSettings = Objects.requireNonNull(applicationSettings);
+        this.metrics = Objects.requireNonNull(metrics);
+        this.vertx = Objects.requireNonNull(vertx);
+        this.timeoutFactory = Objects.requireNonNull(timeoutFactory);
+        this.httpClient = Objects.requireNonNull(httpClient);
+        this.mapper = Objects.requireNonNull(mapper);
+
+        fetchInProgress = new ConcurrentHashSet<>();
+        fetchedData = Caffeine.newBuilder()
+                .<String, AccountFetchContext>build()
+                .asMap();
+    }
+
+    public Future<PriceFloorRules> fetch(Account account) {
+        final PriceFloorRules cachedPriceFloorRules =
+                ObjectUtil.getIfNotNull(fetchedData.get(account.getId()), AccountFetchContext::getRules);
+
+        return cachedPriceFloorRules != null
+                ? Future.succeededFuture(cachedPriceFloorRules)
+                : fetchPriceFloorRules(account);
+    }
+
+    private Future<PriceFloorRules> fetchPriceFloorRules(Account account) {
+        if (account == null) {
+            return Future.succeededFuture();
+        }
+
+        final AccountPriceFloorsFetchConfig fetchConfig = getFetchConfig(account);
+        final Boolean fetchEnabled = ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getEnabled);
+        final String fetchUrl = ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getUrl);
+        final String accountId = account.getId();
+
+        if (!shouldFetchRules(fetchEnabled, fetchUrl, accountId)) {
```

Comment:
This condition can be inverted, so call `fetchPriceFloorRulesAsynchronous(..)` inside.